pySDC/projects/DAE/sweepers/RungeKuttaDAE.py

```diff
@@ -0,0 +1,181 @@
+
+    Note
+    ----
+    When using a RK sweeper to simulate a problem make sure the DAE problem class has a ``du_exact`` method since RK methods need an initial
+    condition for :math:`u'(t_0)` as well.
+
+    In order to implement a new RK method for DAEs a new tableau can be added in ``pySDC.implementations.sweeper_classes.Runge_Kutta.py``.
+    For example, a new method called ``newRungeKuttaMethod`` with nodes :math:`c=(c_1, c_2, c_3)`, weights :math:`b=(b_1, b_2, b_3)` and
+    coefficient matrix
+
+    ..math::
+        \begin{eqnarray}
+            A = \begin{pmatrix}
+                a_{11} & 0 & 0 \\
+                a_{21} & a_{22} & 0 \\
+                a_{31} & a_{32} & & 0 \\
+            \end{pmatrix}
+        \end{eqnarray}
+
+    can be implemented as follows:
+
+    >>> class newRungeKuttaMethod(RungeKutta):
+    >>>     nodes = np.array([c1, c2, c3])
+    >>>     weights = np.array([b1, b2, b3])
+    >>>     matrix = np.zeros((3, 3))
+    >>>     matrix[0, 0] = a11
+    >>>     matrix[1, :2] = [a21, a22]
+    >>>     matrix[2, :] = [a31, a32, a33]
+    >>>     ButcherTableauClass = ButcherTableau
+
+    The new class ``newRungeKuttaMethodDAE`` can then be used by defining the DAE class inheriting from both, this base class and class containing
+    the Butcher tableau:
+
+    >>> class newRungeKuttaMethodDAE(RungeKuttaDAE, newRungeKuttaMethod):
+    >>>     pass
+
+    More details can be found [here](https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/implementations/sweeper_classes/Runge_Kutta.py).
+    """
+
+    def __init__(self, params):
+        super().__init__(params)
+        self.du_init = None
+
+    def predict(self):
+        """
+        Predictor to fill values at nodes before first sweep.
+        """
+
+        # get current level and problem
+        lvl = self.level
+        prob = lvl.prob
+
+        du_init = self.du_init[:] if self.du_init is not None else prob.du_exact(lvl.time)[:]
+        lvl.f[0] = prob.dtype_f(du_init)
+        for m in range(1, self.coll.num_nodes + 1):
+            lvl.u[m] = prob.dtype_u(init=prob.init, val=0.0)
+            lvl.f[m] = prob.dtype_f(init=prob.init, val=0.0)
+
+        lvl.status.unlocked = True
+        lvl.status.updated = True
+
+    def update_nodes(self):
+        r"""
+        Updates the values of solution ``u`` and their gradient stored in ``f``.
+        """
+
+        # get current level and problem description
+        lvl = self.level
+        prob = lvl.prob
+
+        # only if the level has been touched before
+        assert lvl.status.unlocked
+        assert lvl.status.sweep <= 1, "RK schemes are direct solvers. Please perform only 1 iteration!"
+
+        M = self.coll.num_nodes
+        for m in range(M):
+            u_approx = prob.dtype_u(lvl.u[0])
+            for j in range(1, m + 1):
+                u_approx += lvl.dt * self.QI[m + 1, j] * lvl.f[j][:]
+
+            def F(du):
```

Comment:
Is this function shared with the DAE sweeper for SDC? If yes, I suggest having it only once. You can make it a static method of the SDC DAE sweeper and import it here for instance.

Reply:
Having this function only ones avoids some lines. DRY principle I heard from a wise person sometime.. I had an idea setting then
```
def F(u, local_u_approx):
    # do something
``` 
because this function requires the local ``u_approx`` from ``update_nodes``. I also would then make ``u_approx`` make as an argument for the ``solve_system`` method (similar to ``rhs`` in the other sweepers in ``pySDC``). Realising this requires adaption on the other DAE-sweepers and in the parent problem class. Does it probably make sense to do this in another PR, since I know you like clean PR's?